miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+
+        array = MemArray(vm, addr, Num("I"))
+        array[2] = 5
+        array[4:8] = [0, 1, 2, 3]
+        print array[20]
+
+    If the @field_type is a Ptr, deref_get(index) and deref_set(index) can be
+    used to dereference a field at a given index in the array.
+
+    mem_array_type can be used to generate a type that includes the field_type.
+    Such a generated type can be instanciated with only vm and addr, as are
+    other MemStructs.
+    """
+    _field_type = None
+
+    def __init__(self, vm, addr=None, field_type=None):
+        if self._field_type is None:
+            self._field_type = field_type
+        if self._field_type is None:
+            raise NotImplementedError(
+                "Provide field_type to instanciate this class, "
+                "or generate a subclass with mem_array_type.")
+        super(MemArray, self).__init__(vm, addr)
+
+    @property
+    def field_type(self):
+        """Return the MemField subclass instance that represents the type of
+        this MemArray items.
+        """
+        return self._field_type
+
+    def _normalize_idx(self, idx):
+        # Noop for this type
+        return idx
+
+    def _normalize_slice(self, slice_):
+        start = slice_.start if slice_.start is not None else 0
+        stop = slice_.stop if slice_.stop is not None else self.get_size()
+        step = slice_.step if slice_.step is not None else 1
+        return slice(start, stop, step)
+
+    def _check_bounds(self, idx):
+        idx = self._normalize_idx(idx)
+        if not isinstance(idx, int) and not isinstance(idx, long):
+            raise ValueError("index must be an int or a long")
+        if idx < 0:
+            raise IndexError("Index %s out of bounds" % idx)
+
+    def index2addr(self, idx):
+        """Return the address corresponding to a given @index in this MemArray.
+        """
+        self._check_bounds(idx)
+        addr = self.get_addr() + idx * self._field_type.size()
+        return addr
+
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
+            res = []
+            idx = self._normalize_slice(idx)
+            for i in xrange(idx.start, idx.stop, idx.step):
+                res.append(self._field_type.get(self._vm, self.index2addr(i)))
+            return res
+        else:
+            return self._field_type.get(self._vm, self.index2addr(idx))
+
+    def deref_get(self, idx):
+        """If self.field_type is a Ptr, return the MemStruct self[idx] points
+        to.
+        """
+        return self._field_type.deref_get(self._vm, self[idx])
+
+    def __setitem__(self, idx, item):
+        if isinstance(idx, slice):
+            idx = self._normalize_slice(idx)
+            if len(item) != len(xrange(idx.start, idx.stop, idx.step)):
+                raise ValueError("Mismatched lengths in slice assignment")
+            for i, val in zip(xrange(idx.start, idx.stop, idx.step), item):
```

Comment:
If you want to fully use iterators, you can use `itertools.izip` instead of `zip`